File: jihoon/Coding_test (1)/Algorithms/Sorting_alg/LL_mergesort/LL_mergesort.c
#include <stdio.h>
#include <stdlib.h>

typedef struct LinkedNode {
    int val;
    struct LinkedNode* next;
} LinkedNode;

LinkedNode* merge_sort(LinkedNode* head);
LinkedNode* merge(LinkedNode* left, LinkedNode* right);
LinkedNode* list_to_linked_list(int arr[], int n);
void print_list(LinkedNode* head);
void free_list(LinkedNode* head);

int main() {
    int a[] = {1, 3, 2, 4, 5, 6, 7, 8};
    int b[] = {5, 3, 2, 1, 4, 0, 6, 7};

    LinkedNode* ahead = list_to_linked_list(a, sizeof(a) / sizeof(a[0]));
    LinkedNode* bhead = list_to_linked_list(b, sizeof(b) / sizeof(b[0]));

    LinkedNode* sorted_ahead = merge_sort(ahead);
    LinkedNode* sorted_bhead = merge_sort(bhead);

    print_list(sorted_ahead);
    print_list(sorted_bhead);
    
    free_list(sorted_ahead);
    free_list(sorted_bhead);

    return 0;
}

LinkedNode* merge_sort(LinkedNode* head) {
    if (!head || !head->next) {
        return head;
    }

    LinkedNode *prev = NULL, *slow = head, *fast = head;
    while (fast && fast->next) {
        prev = slow;
        slow = slow->next;
        fast = fast->next->next;
    }
    prev->next = NULL;

    LinkedNode* left = merge_sort(head);
    LinkedNode* right = merge_sort(slow);

    return merge(left, right);
}

LinkedNode* merge(LinkedNode* left, LinkedNode* right) {
    if (!left) return right;
    if (!right) return left;

    if (left->val > right->val) {
        left->next = merge(left->next, right);
        return left;
    } else {
        right->next = merge(left, right->next);
        return right;
    }
}

LinkedNode* list_to_linked_list(int arr[], int n) {
    if (n <= 0) return NULL;
    LinkedNode* head = (LinkedNode*)malloc(sizeof(LinkedNode));
    head->val = arr[0];
    head->next = NULL;

    LinkedNode* current = head;
    for (int i = 1; i < n; ++i) {
        current->next = (LinkedNode*)malloc(sizeof(LinkedNode));
        current = current->next;
        current->val = arr[i];
        current->next = NULL;
    }

    return head;
}

void print_list(LinkedNode* head) {
    while (head) {
        printf("%d -> ", head->val);
        head = head->next;
    }
    printf("NULL\n");
}

void free_list(LinkedNode* head) {
    while (head) {
        LinkedNode* temp = head;
        head = head->next;
        free(temp);
    }
}

File: jihoon/Coding_test (1)/Algorithms/Sorting_alg/LL_mergesort/LL_mergesort.py
# linked list node 구현
class LinkedNode:
    def __init__(self, val):
        self.val = val
        self.next = None

# linked list를 이용한 in-place merge sort 구현
def merge_sort(head: LinkedNode) -> LinkedNode:
    if not head or not head.next:
        return head
    
    # linked list를 두 개의 sub list로 나누기
    prev, slow, fast = None, head, head
    while fast and fast.next:
        prev = slow
        slow = slow.next
        fast = fast.next.next
    prev.next = None
    
    # 재귀적으로 sub list를 정렬하고 병합
    left = merge_sort(head)
    right = merge_sort(slow)
    return merge(left, right)

# 두 개의 linked list를 병합하는 함수
def merge(left: LinkedNode, right: LinkedNode) -> LinkedNode:
    if not left:
        return right
    if not right:
        return left
    
    if left.val < right.val: # 부등호 방향 < : asc,  > : desc
        left.next = merge(left.next, right)
        return left
    else:
        right.next = merge(left, right.next)
        return right

def list_to_linked_list(numlist:list)->LinkedNode:
    if not numlist:
        return None
    head = LinkedNode(numlist[0])
    curr = head

    for val in numlist[1:]:
        curr.next = LinkedNode(val)
        curr = curr.next
    return head

def printnode(head: LinkedNode)->None:
    while(head):
        print(head.val, end = "->")
        head = head.next
    return


if __name__ == "__main__":
    a = [1, 3, 2, 4, 5, 6, 7, 8]; ahead = list_to_linked_list(a)
    b = [5, 3, 2, 1, 4, 0, 6, 7]; bhead = list_to_linked_list(b)
    # printnode(ahead)
    # printnode(bhead)
    printnode(merge_sort(ahead))
    printnode(merge_sort(bhead))